### PR TITLE
New features for CodeParrot training script

### PR DESCRIPTION
**Pull request by loubnabnl**
This PR adds some features to CodeParrot training script.

- Add TFLOPS to logging
- Use Accelerate checkpointing and tracking for Wandb and Tensorborad
- Fix gradient accumulation for DDP (https://github.com/huggingface/accelerate/pull/106)
- Scale loss approprietly for the last batch
- Fix typo in the README

cc @lvwerra @LysandreJik

### CODE REVIEW

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 1]
```diff
@@ -140,7 +137,10 @@ def evaluate(args):
         losses.append(accelerator.gather(loss))
         if args.max_eval_steps > 0 and step >= args.max_eval_steps:
             break
-    loss = torch.mean(torch.cat(losses))
+    losses = torch.cat(losses)
+    eval_size = eval_dataloader.dataset.current_size
+    losses = losses[:eval_size]
+    loss = torch.mean(losses)
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:143]: Maybe we could wrap that into a single line:
```suggestion
    loss = losses[:eval_dataloader.dataset.current_size].mean()
```

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 2]
```diff
@@ -203,26 +203,52 @@ def get_lr():
...
+        batch_size = args.train_batch_size * accelerator.state.num_processes * args.gradient_accumulation_steps
+        factor = (
+            24 * checkpoint_factor * batch_size * args.seq_length * config_model.n_layer * (config_model.n_embd**2)
+        )
+        flops_per_iteration = factor * (
+            1.0
+            + (args.seq_length / (6.0 * config_model.n_embd))
+            + (tokenizer.vocab_size / (16.0 * config_model.n_layer * config_model.n_embd))
+        )
+        tflops = flops_per_iteration / (elapsed_time_per_iteration * accelerator.state.num_processes * (10**12))
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:239]: could we move that to a dedicated function? e.g. `compute_tflops(elapsed_time, accelerator, args)`? It would be nice if the main training loop would stay concise to make it clearer what's going on.

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 3]
```diff
@@ -203,26 +203,52 @@ def get_lr():
...
+        )
+        t_start = time.time()
     if step % args.save_checkpoint_steps == 0:
         logger.info("Evaluating and saving model checkpoint")
         eval_loss, perplexity = evaluate(args)
         log_metrics(step, {"loss/eval": eval_loss, "perplexity": perplexity})
         accelerator.wait_for_everyone()
         unwrapped_model = accelerator.unwrap_model(model)
         unwrapped_model.save_pretrained(args.save_dir, save_function=accelerator.save)
+        accelerator.save_state(args.save_dir)
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:251]: What exactly does the `save_state` save? A bunch of files? Maybe we could add them to a folder e.g. `args.save_dir + "/state/"`.

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 4]
```diff
@@ -236,5 +262,6 @@ def get_lr():
 accelerator.wait_for_everyone()
 unwrapped_model = accelerator.unwrap_model(model)
 unwrapped_model.save_pretrained(args.save_dir, save_function=accelerator.save)
+accelerator.save_state(args.save_dir)
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:265]: Same as above

**Review by lvwerra — COMMENTED** [review]
Thanks @loubnabnl, looks pretty clean already! I left a few minor comments, mainly to make the code a bit more concise.

Regarding the saving of the state: I think it is great that we can now save it but I think what is missing currently is the mechanism to start the script from a saved state. I don't think we need to do much and you can probably follow the example here:
https://github.com/huggingface/accelerate/blob/main/examples/complete_nlp_example.py

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 3 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py]: `save_state` returns a bunch of files (model, optimizer ..), I'm now saving them in folders corresponding to the steps to be able to resume training from these steps later

Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py]: and since `save_state` already saves the model in the folder step, I now use save_pretrained for the unwrapped model only for the last checkpoint to save model in args.save_dir to load direclty from there later

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 5]
```diff
@@ -126,8 +125,21 @@ def get_grouped_params(model, args, no_decay=["bias", "LayerNorm.weight"]):
 def log_metrics(step, metrics):
     logger.info(f"Step {step}: {metrics}")
     if accelerator.is_main_process:
-        wandb.log(metrics)
-        [tb_writer.add_scalar(k, v, step) for k, v in metrics.items()]
+        accelerator.log(metrics, step)
+
+
+def compute_tflops(elapsed_time, accelerator, args):
+    config_model = accelerator.unwrap_model(model).config
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:132]: Minor thing: can you add the link to the formula here? either BigScience or the paper itself. So somebody could find out where that black magic formula actually comes from :)

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 6]
```diff
@@ -200,29 +214,60 @@ def get_lr():
...
+
 # Train model
 model.train()
 completed_steps = 0
+t_start = time.time()
 for step, batch in enumerate(train_dataloader, start=1):
+    if args.resume_from_checkpoint:
+        # We need to skip steps until we reach the resumed step
+        if step < resume_step:
+            continue
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:240]:
```suggestion
    if args.resume_from_checkpoint and step < resume_step:
            continue # we need to skip steps until we reach the resumed step
```

**Review by lvwerra — APPROVED** [review]
Only two minor comments and then it is good to go! 🚀

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 5 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:133]: Done!
